src/python/pants/build_graph/address.py

```diff
@@ -383,11 +384,22 @@ def __init__(
         :param relative_file_path: The relative path from the spec_path to an addressed file,
           if any. Because files must always be located below targets that apply metadata to
           them, this will always be relative.
+        :param _original_address: Used internally by the Address class when creating a generator
+          address for the original target address in order to support going back
+          (lossless conversion).
         """
+        # Support providing a relative_file_path in the target name as "<filename>:<target_name>".
+        if target_name and not relative_file_path and ":" in target_name:
+            relative_file_path, target_name = target_name.rsplit(":", 1)
+            self._is_explicit_file_target = True
+        else:
+            self._is_explicit_file_target = False
+
         self.spec_path = spec_path
         self.parameters = FrozenDict(parameters) if parameters else FrozenDict()
         self.generated_name = generated_name
         self._relative_file_path = relative_file_path
+        self._original_address = _original_address
```

Comment:
Will there be a performance hit/memory penalty to introducing these extra slots on each address?
Would it make a difference if we declared a `__slots__` attribute on the class?